Non-plugin version of hsmtool

darosior opened the pull request:
This closes #3177 as it is a concurrent version of the `hsmtools` functionalities.

Changelog-None

niftynei commented on tests/test_wallet.py:
```diff
@@ -606,3 +605,61 @@ def test_hsm_secret_encryption(node_factory, executor):
+    l1.daemon.start(stdin=slave_fd, stderr=subprocess.STDOUT,
+                    wait_for_initialized=False)
+    time.sleep(3 if SLOW_MACHINE else 1)
+    os.write(master_fd, "{}".format(password).encode("utf-8"))
```
```suggestion
    os.write(master_fd, password.encode("utf-8"))
```

darosior replied:
Hmmm I might have wrote this while there was an interesting talk ongoing :joy:

niftynei commented on tests/test_wallet.py:
```diff
@@ -606,3 +605,61 @@ def test_hsm_secret_encryption(node_factory, executor):
+@unittest.skipIf(VALGRIND, "It does not play well with prompt and key derivation.")
+def test_hsmtools_secret_decryption(node_factory):
+    l1 = node_factory.get_node()
+    password = "reckless\n"
```
```suggestion
    password = "reckless\n".encode("utf-8") 
```

?? then you can use it inline elsewhere.

darosior replied:
Yeah but then I would have to decode it for Pylightning `call()` method, which (since we also need to remove the trailing `\n`) is arguably dirtier ?

niftynei commented on tests/test_wallet.py:
```diff
@@ -606,3 +605,61 @@ def test_hsm_secret_encryption(node_factory, executor):
+    l1.daemon.opts.pop("encrypted-hsm")
+    l1.daemon.start(stdin=slave_fd, wait_for_initialized=True)
+    assert node_id == l1.rpc.getinfo()["id"]
+    l1.stop()
```
should we also verify that an incorrect password fails to decrypt?

darosior replied:
Absolutely ! Thanks

niftynei commented on tools/hsmtools.c:
```diff
@@ -0,0 +1,214 @@
+	close(fd);
+
+	/* Create a backup file, "just in case". */
+	rename(hsm_secret_path, "hsm_secret.backup");
```
does `rename` update just the file name, or the entire path?

darosior replied:
entire path

niftynei replied:
a bit of a stretch goal, but it'd be nice if it's in the same directory as the original hsm file, so that it's easy to find/recover in case of a problem

darosior replied:
Actually it was in plugin version (we chdir in libplugin.c at `init`) but I kept it although we don"t run it from the same directory.
In case of a not-already handled problem, the backup is still in the running directory.. And it would be quite hacky, just for a tool  : recovering the path to the lightning-dir from the hsm_secret path, then prefixing the rename..

niftynei commented on tools/hsmtools.c:
```diff
@@ -0,0 +1,214 @@
+		rename("hsm_secret.backup", hsm_secret_path);
+		errx(ERROR_HSM_FILE, "Failure writing cipher to hsm_secret.");
+	}
+	if (!write_all(fd, cipher, sizeof(cipher))) {
```
i don't think you need this, you're writing the cipher in the previous if statement as well.

darosior replied:
I need to write both the header and the body of the cipher : the variable name is maybe not very explicit..

niftynei replied:
yeah you write the cipher already in L161

darosior replied:
Oh yes, I forgot to remove it here thanks

darosior commented on tests/test_wallet.py:
```diff
@@ -606,3 +605,61 @@ def test_hsm_secret_encryption(node_factory, executor):
+    node_id = l1.rpc.getinfo()["id"]
+    l1.stop()
+
+    # Decrypt it via RPC...
```
Oops that's a trailing comment from the plugin version :-)

darosior commented on tests/test_wallet.py:
```diff
@@ -606,3 +605,61 @@ def test_hsm_secret_encryption(node_factory, executor):
+    l1.stop()
+
+    # Decrypt it via RPC...
+    cmd_line = "tools/hsmtools decrypthsm {} {}".format(hsm_path, password[:-1])
```
I find it more readable this way, but I can make it a list

darosior commented:
Thanks for the review, I am about to push the polish and the `dumpcommitments` method. I'll integrate your comments.

darosior commented on tests/test_wallet.py:
```diff
@@ -606,3 +605,61 @@ def test_hsm_secret_encryption(node_factory, executor):
+    l1.stop()
+    l1.daemon.start(stdin=slave_fd, stderr=subprocess.STDOUT,
+                    wait_for_initialized=False)
+    time.sleep(3 if SLOW_MACHINE else 1)
```
I can't as with `wait_for_log` the log messages is already in the log when we start watching. I at first used an executor but `is_in_logs` is much neater

Sjors commented:
Nice! I prefer `Co-Authored-By: Sjors Provoost <sjors@sprovoost.nl>` over an @ mention in the commit, because the latter can lead to lots of altcoin spam... (when they get around to cloning this)

darosior commented:
Great, Github even takes it into account.

darosior commented:
> Can you add a (optional) node public key argument to dumpcommitments? I don't know if this gets purged from the database 30 (?) days after channel closure, but regardless it's nice to be able to dump a commitment without having access to the full database.

You don't need access to the database to dump the commitment, just the `hsm_secret` from which we derive the node pubkey : that's why I removed it, it's redundant if you already have the hsm

darosior commented:
> Can you add a (optional) node public key argument to dumpcommitments? I don't know if this gets purged from the database 30 (?) days after channel closure, but regardless it's nice to be able to dump a commitment without having access to the full database.

Yes, you right this should be the default I think.

Sjors commented:
I'm able to produce the same commitment point with this PR as with my original tool.

darosior commented:
Polished, corrected, and rebased on master.

rustyrussell commented on tools/hsmtools.c:
```diff
@@ -0,0 +1,135 @@
+	if (close(fd) != 0)
+		return false;
+
+	fd = open(".", O_RDONLY);
```
This is fsyncing the wrong dir?  You want dirname(hsm_secret_path) here.

darosior replied:
Bad copy/paste!

rustyrussell commented on tools/hsmtools.c:
```diff
@@ -19,6 +19,7 @@ static void show_usage(void)
 	printf("./hsmtools <method> [arguments]\n");
 	printf("methods:\n");
 	printf("	- decrypthsm <path/to/hsm_secret> <password>\n");
+	printf("	- encrypthsm <path/to/hsm_secret> <password>\n");
```
Since the tool is call htmtools, we could just call these commands "decrypt" and "encrypt" perhaps?  And maybe "hsmtool" is better than "hsmtools" though that's marginal.

darosior replied:
Yeah I wanted to add more tool*s* (https://github.com/ElementsProject/lightning/pull/3217) but I can make it singular ^^

rustyrussell commented on tools/hsmtools.c:
```diff
@@ -44,25 +53,33 @@ static bool ensure_hsm_secret_exists(int fd)
+
+	fd = open(hsm_secret_path, O_RDONLY);
+	if (fd < 0)
+		errx(ERROR_HSM_FILE, "Could not open hsm_secret");
```
Use err() not errx() here, which will tell them what the error was?

darosior replied:
err.. I confused the err functions

rustyrussell commented on tools/hsmtools.c:
```diff
@@ -44,25 +53,33 @@ static bool ensure_hsm_secret_exists(int fd)
+	if (fd < 0)
+		errx(ERROR_HSM_FILE, "Could not open hsm_secret");
+	if (!read_all(fd, hsm_secret, sizeof(*hsm_secret)))
+		errx(ERROR_HSM_FILE, "Could not read hsm_secret");
```
Here too...

darosior commented:
Corrected the `fsync` and the confusion between `err` functions, renamed encryption methods to `encrypt` and `decrypt`, and renamed the tool to `hsmtool`. Thanks for the review!
